Document IndexDB

rtm0 opened the pull request:
### Describe Your Changes

This is an attempt to document IndexDB. I guess I was trying to touch the important points that might be of interest for the end users while refraining from making it too detailed (such as I did not enumerate and describe all the specific record types).

Please take a look and any suggestions are very welcome.

### Checklist

The following checks are **mandatory**:

- [x ] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

rtm0 commented:
I thought I'd add `Phuong Le` as the reviewer for his experience in writing technical docs but could not find his username in GitHub. If you know his username, please add him. Thanks!

qphuongl commented on docs/README.md:
```diff
@@ -1905,7 +1905,55 @@ See more details in [monitoring docs](#monitoring).
+the raw samples by metric name, label name, and label value by mapping these
+values to the corresponding TSIDs.
+
+There are two types of inverted index: `global` and `per-day`. Initially,
+VictoriaMetrics had only global index. The per-day index has been added later as
+an optimization that speeds up data retrieval when
+[high churn rate](https://docs.victoriametrics.com/faq/#what-is-high-churn-rate)
+is the case.
+
+VictoriaMetrics decides which index to use based on the length of the search
+time range. If it is less than or equal to 40 days, then the per-day index is
+used. Otherwise, it will use the global index.
```
I suggest using different formats instead of long text in this case, for your reference:

---
VictoriaMetrics uses two types of inverted indexes:
- Global Index: This was the original type of index used by VictoriaMetrics. It applies across the entire retention period, meaning that each MetricName -> TSID record is created only once per retention period.
- Per-day Index: This index was introduced later as an optimization for handling [high churn rates](https://docs.victoriametrics.com/faq/#what-is-high-churn-rate) (frequent creation and deletion of time series). The per-day index stores mappings similar to the global index but includes the date in each record, which speeds up data retrieval for queries within a shorter time range (up to 40 days).

When a search query is executed, VictoriaMetrics decides which index to use based on the time range of the query:
- Per-day Index is used if the search time range is 40 days or less.
- Global Index is used for search queries with a time range greater than 40 days.

qphuongl reviewed (commented):
Looks good and coherent, I only suggest improving the readability of the text.

rtm0 commented:
Thanks @qphuongl! I've changed the formatting and also removed the example as it had been introduced only to describe the difference in how mappings are created in both indexes. Please take a look.

hagen1778 commented on docs/README.md:
```diff
@@ -1905,7 +1905,54 @@ See more details in [monitoring docs](#monitoring).
+
+VictoriaMetrics uses two types of inverted indexes:
+
+-   Global index. This was the original type of index used by VictoriaMetrics.
```
```suggestion
-   Global index.
```
I don't see much value for a reader to mention this.

rtm0 replied:
Removed.

hagen1778 commented on docs/README.md:
```diff
@@ -1905,7 +1905,54 @@ See more details in [monitoring docs](#monitoring).
+
+-   Global index. This was the original type of index used by VictoriaMetrics.
+    Searches using this index is performed across the entire retention period.
+-   Per-day index. This index was introduced later as an optimization for
+    handling.
+    [high churn rate](https://docs.victoriametrics.com/faq/#what-is-high-churn-rate)
+    (frequent creation and deletion of time series). The per-day index stores
```
```suggestion
-   Per-day index. The per-day index stores
```

rtm0 replied:
Removed.

hagen1778 commented on docs/README.md:
```diff
@@ -1905,7 +1905,54 @@ See more details in [monitoring docs](#monitoring).
+-   In the per-day index each mapping will be created for each unique date that
+    has been seen in the samples for the corresponding time series.
+
+NOTE: The mappings are not created atomically and duplicates are possible when
```
I don't see value in this sentence. It should be either removed or linked to some behavior that user can experience or alter. In other words, it has no practical knowledge for a user.

rtm0 replied:
Removed.

hagen1778 reviewed (commented):
Thanks for the PR!
What's the purpose of this documentation? Is there a way user can use this doc somehow?
I'm not against documenting stuff, but from my perspective each doc section should be actionable somehow for a user. For example, the TSID thing could have been linked to metrics or logs related to registering these IDs.
Do you see other value (besides pure curiosity) in this addition?

rtm0 commented:
> Do you see other value (besides pure curiosity) in this addition?

So I've been working on a feature of optionally disabling per-day index using a flag (no PR yet, please see the [branch commits in my repo](https://github.com/VictoriaMetrics/VictoriaMetrics/compare/master...rtm0:VictoriaMetrics:disable-per-date-index). The work also includes the [documentation](https://github.com/rtm0/VictoriaMetrics/blob/disable-per-date-index/docs/README.md?plain=1#L2485-L2539) of why the flag is needed, how to use it and what to expect. The documentation mentions IndexDB. But there is no documentation about it anywhere. And so I've decided to write one to the background info on IndexDB.

hagen1778 reviewed (approved):
LGTM
